bundles/org.openhab.voice.porcupineks/src/main/java/org/openhab/voice/porcupineks/internal/PorcupineKSService.java

```diff
@@ -0,0 +1,283 @@
+import org.openhab.core.voice.KSpottedEvent;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ai.picovoice.porcupine.Porcupine;
+import ai.picovoice.porcupine.PorcupineException;
+
+/**
+ * The {@link PorcupineKSService} is a keyword spotting implementation based on porcupine.
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME, description_uri = SERVICE_CATEGORY + ":"
+        + SERVICE_ID)
+public class PorcupineKSService implements KSService {
+    private final Logger logger = LoggerFactory.getLogger(PorcupineKSService.class);
+    private static final String PORCUPINE_FOLDER = Path.of(OpenHAB.getUserDataFolder(), ".porcupine").toString();
+    private static final String EXTRACTION_FOLDER = Path.of(OpenHAB.getUserDataFolder(), ".porcupine", "extracted")
+            .toString();
+    private final ScheduledExecutorService executor = ThreadPoolManager.getScheduledPool("audio");
+    private PorcupineKSConfiguration config = new PorcupineKSConfiguration();
+    private boolean loop = false;
+    private @Nullable BundleContext bundleContext;
+    private @Nullable Porcupine porcupine = null;
+    private @Nullable Future<?> scheduledTask;
+
+    static {
+        var logger = LoggerFactory.getLogger(PorcupineKSService.class);
+        File directory = new File(PORCUPINE_FOLDER);
+        if (!directory.exists()) {
+            if (directory.mkdir()) {
+                logger.info("porcupine dir created {}", PORCUPINE_FOLDER);
+            }
+        }
+        File childDirectory = new File(EXTRACTION_FOLDER);
+        if (!childDirectory.exists()) {
+            if (childDirectory.mkdir()) {
+                logger.info("porcupine extraction file dir created {}", EXTRACTION_FOLDER);
+            }
+        }
+    }
+
+    @Activate
+    protected void activate(ComponentContext componentContext, Map<String, Object> config) {
+        var serviceConfig = new org.openhab.core.config.core.Configuration(config).as(PorcupineKSConfiguration.class);
+        this.config = serviceConfig;
+        this.bundleContext = componentContext.getBundleContext();
+        if (serviceConfig.apiKey.isBlank()) {
+            logger.warn("Missing pico voice api key to use Porcupine Keyword Spotter");
+        }
+    }
+
+    private String prepareLib(BundleContext bundleContext, String path) throws IOException {
+        var relativePath = path.substring(path.indexOf("porcupine/"));
+        var porcupineResource = bundleContext.getBundle().getEntry(relativePath);
```

Comment:
This method extract the binary files from the bundle to the extracted folder and returns an absolute path to the extracted file, because the underlying binary need the file to exists in order to work.
As a summary I'm relying on the internal porcupine utils to get the correct library path, but it resolves paths that do not exists as files are inside the bundle, this method use those path to locale a file inside the bundle, extract it, and return an usable path.
Hope I explained it correctly, let me know if not.

Reply:
It could become clearer for me if you tell what are the values for `Porcupine.LIBRARY_PATH` and `Porcupine.MODEL_PATH`.